packages/celotool/src/e2e-tests/slashing_tests.ts

```diff
@@ -0,0 +1,249 @@
+      )
+      const hash = await contract.methods.hashHeader(header).call()
+      assert.equal(hash, '0xf5a450266c77dce47f7698959d8e7019db860ee19a5322b16a853fdf23607100')
+    })
+
+    it('should hash current header correctly', async () => {
+      const contract = await kit._web3Contracts.getElection()
+      const current = await kit.web3.eth.getBlockNumber()
+      const block = await kit.web3.eth.getBlock(current)
+      const rlpEncodedBlock = rlp.encode(headerArray(kit.web3, block))
+      const blockHash = await contract.methods.hashHeader(rlpEncodedBlock).call()
+      assert.equal(blockHash, block.hash)
+    })
+  })
+
+  let doubleSigningBlock: any
+
+  describe('test slashing for downtime', () => {
+    before(async function(this: any) {
+      this.timeout(0) // Disable test timeout
+      await restartWithDowntime()
+    })
+
+    it('slash for downtime', async function(this: any) {
+      this.timeout(0) // Disable test timeout
+      const slasher = await kit._web3Contracts.getDowntimeSlasher()
+      const blockNumber = await web3.eth.getBlockNumber()
+      await waitUntilBlock(blockNumber + 20)
+      doubleSigningBlock = await web3.eth.getBlock(blockNumber + 15)
+
+      const signer = await slasher.methods.validatorSignerAddressFromSet(4, blockNumber + 12).call()
+
+      const validator = (await kit.web3.eth.getAccounts())[0]
+      await kit.web3.eth.personal.unlockAccount(validator, '', 1000000)
+      const lockedGold = await kit.contracts.getLockedGold()
+
+      const validatorsContract = await kit._web3Contracts.getValidators()
+      const history = await validatorsContract.methods.getMembershipHistory(signer).call()
+      const historyIndex = history[0].length - 1
+
+      await slasher.methods
+        .slash(
+          blockNumber + 12,
+          4,
+          4,
+          historyIndex,
+          [],
+          [],
+          [],
+          [NULL_ADDRESS],
+          [NULL_ADDRESS],
+          [0]
+        )
+        .send({ from: validator, gas: 5000000 })
+
+      const balance = await lockedGold.getAccountTotalLockedGold(signer)
+      assert.equal(balance.toString(10), '9000000000000000000000')
+    })
+  })
+
+  describe('test slashing for double signing', () => {
+    before(async function(this: any) {
+      this.timeout(0) // Disable test timeout
+      await restart()
+    })
+
+    it('slash for double signing', async function(this: any) {
+      this.timeout(0) // Disable test timeout
+      const slasher = await kit._web3Contracts.getDoubleSigningSlasher()
+
+      await waitUntilBlock(doubleSigningBlock.number)
+
+      const other = headerFromBlock(web3, doubleSigningBlock)
+
+      const num = await slasher.methods.getBlockNumberFromHeader(other).call()
+
+      const header = headerFromBlock(web3, await web3.eth.getBlock(num))
+
+      const bitmap1 = await slasher.methods.getVerifiedSealBitmapFromHeader(header).call()
```

Comment:
nit: Please add a comment above this, e.g. "Find a validator that double signed. Both blocks will have signatures from exactly 2F+1 validators."